Lib/test/test_marshal.py

```diff
@@ -344,6 +344,20 @@ def test_eof(self):
         for i in range(len(data)):
             self.assertRaises(EOFError, marshal.loads, data[0: i])
 
+    def test_deterministic_sets(self):
+        # bpo-37596: sets and frozensets must always marshal deterministically!
+        # Test this by seeding string hashes:
+        elements = "(float('nan'), b'a', b'b', b'c', 'x', 'y', 'z')"
+        for kind in ("set", "frozenset"):
+            script = f"import marshal; print(marshal.dumps({kind}({elements})))"
+            with self.subTest(kind):
+                # {nan, b'b', 'x', b'a', b'c', 'y', 'z'}
```

Comment:
I suggest adding Serhiy's test case as well, ``{("string", 1), ("string", 2), ("string", 3)}``.  If someone later tries to "simplify" this algorithm, the test would provide early detection of the backreference problem.